### PR TITLE
mds: initialize epoch for quiescedb

### PR DESCRIPTION
**Pull request by batrick**
Fixes: https://tracker.ceph.com/issues/66449

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
- Component impact
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66462](https://tracker.ceph.com/issues/66462).

**Review by leonid-s-usov — APPROVED** [review]
Thank you Patrick!

**Comment by batrick** [comment]
It seems there is a follow-up issue but I'm not immediately sure why this wasn't also solved by this patch:

```
<error>
  <unique>0x42d</unique>
  <tid>29</tid>
  <threadname>quiesce_db_mgr</threadname>
  <kind>UninitCondition</kind>
  <what>Conditional jump or move depends on uninitialised value(s)</what>
  <stack>
    <frame>
      <ip>0x5E732A</ip>
      <obj>/usr/bin/ceph-mds</obj>
      <fn>UnknownInlinedFun</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/mds</dir>
      <file>QuiesceDb.h</file>
      <line>129</line>
    </frame>
    <frame>
      <ip>0x5E732A</ip>
      <obj>/usr/bin/ceph-mds</obj>
      <fn>QuiesceDbManager::leader_upkeep_db()</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/mds</dir>
      <file>QuiesceDbManager.cc</file>
      <line>889</line>
    </frame>
    <frame>
      <ip>0x5D923F</ip>
      <obj>/usr/bin/ceph-mds</obj>
      <fn>QuiesceDbManager::leader_upkeep(std::deque&lt;QuiesceDbPeerAck, std::allocator&lt;QuiesceDbPeerAck&gt; &gt;&amp;&amp;, std::deque&lt;QuiesceDbManager::RequestContext*, std::allocator&lt;QuiesceDbManager::RequestContext*&gt; &gt;&amp;&amp;)</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/mds</dir>
      <file>QuiesceDbManager.cc</file>
      <line>423</line>
    </frame>
    <frame>
      <ip>0x5DC856</ip>
      <obj>/usr/bin/ceph-mds</obj>
      <fn>QuiesceDbManager::quiesce_db_thread_main()</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/mds</dir>
      <file>QuiesceDbManager.cc</file>
      <line>111</line>
    </frame>
    <frame>
      <ip>0x210C80</ip>
      <obj>/usr/bin/ceph-mds</obj>
      <fn>QuiesceDbManager::QuiesceDbThread::entry()</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/mds</dir>
      <file>QuiesceDbManager.h</file>
      <line>217</line>
    </frame>
    <frame>
      <ip>0x4C89D14</ip>
      <obj>/usr/lib64/ceph/libceph-common.so.2</obj>
      <fn>Thread::entry_wrapper()</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/common</dir>
      <file>Thread.cc</file>
      <line>87</line>
    </frame>
    <frame>
      <ip>0x4C89D30</ip>
      <obj>/usr/lib64/ceph/libceph-common.so.2</obj>
      <fn>Thread::_entry_func(void*)</fn>
      <dir>/usr/src/debug/ceph-19.0.0-4302.g119deb3b.el9.x86_64/src/common</dir>
      <file>Thread.cc</file>
      <line>74</line>
    </frame>
    <frame>
      <ip>0x5952C01</ip>
      <obj>/usr/lib64/libc.so.6</obj>
      <fn>start_thread</fn>
    </frame>
    <frame>
      <ip>0x59D6F33</ip>
      <obj>/usr/lib64/libc.so.6</obj>
      <fn>clone</fn>
    </frame>
  </stack>
</error>
```

Or I no longer reproduced it locally not because of this fix and the problem is somewhere else.

File: src/mds/QuiesceDbManager.h [hunk 1]
```diff
@@ -250,7 +250,7 @@ class QuiesceDbManager {
     struct PeerInfo {
         QuiesceMap diff_map;
         QuiesceTimePoint last_activity;
-        QuiesceDbVersion last_sent_version;
+        QuiesceDbVersion last_sent_version{};
```
Comment (batrick) [inline src/mds/QuiesceDbManager.h:253]: @leonid-s-usov this was missed too. Please re-approve if you think this is good.

Comment (leonid-s-usov) [reply src/mds/QuiesceDbManager.h:253]: I would prefer to have the change at the QuiesceDbVersion struct definition site. All we need to do is to provide initializers to the epoch and the set_version.

```
struct QuiesceDbVersion {
  epoch_t epoch = 0;
  QuiesceSetVersion set_version = 0;
  auto operator<=>(QuiesceDbVersion const& other) const = default;
  QuiesceDbVersion& operator+(unsigned int delta) {
    set_version += delta;
    return *this;
  }
};
```

With this approach, people using the type won't have to worry about initializing it explicitly.

**Review by leonid-s-usov — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66609](https://tracker.ceph.com/issues/66609).

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/66609#note-1

**Pull request merged**